collections/collections.go

```diff
@@ -0,0 +1,90 @@
+package collections
+
+import (
+	"errors"
+	"math"
+
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+)
+
+var (
+	// ErrNotFound is returned when the provided key is not present in the StorageProvider.
+	ErrNotFound = errors.New("collections: not found")
+	// ErrEncoding is returned when something fails during key or value encoding/decoding.
+	ErrEncoding = errors.New("collections: encoding error")
+)
+
+// StorageProvider represents sdk.Context
+// it is used to avoid to reduce dependencies.
+type StorageProvider interface {
+	// KVStore returns a KVStore given its StoreKey.
+	KVStore(key storetypes.StoreKey) storetypes.KVStore
+}
+
+// Prefix defines a segregation namespace
+// for specific collections objects.
+type Prefix struct {
+	raw []byte // TODO(testinginprod): maybe add a humanized prefix?
+}
+
+// Bytes returns the raw Prefix bytes.
+func (n Prefix) Bytes() []byte { return n.raw }
+
+// NewPrefix returns a Prefix given the provided namespace identifier.
+// Prefixes of the same module must not start
+func NewPrefix[T interface{ int | string | []byte }](identifier T) Prefix {
+	i := any(identifier)
+	var prefix []byte
+	switch c := i.(type) {
+	case int:
+		if c > math.MaxUint8 || c < 0 {
+			panic("invalid integer prefix value: must be between 0 and 255")
+		}
+		prefix = []byte{uint8(c)}
+	case string:
+		prefix = []byte(c)
+	case []byte:
+		prefix = c // maybe copy?
+	}
+	return Prefix{raw: prefix}
+}
+
+// KeyCodec defines a generic interface which is implemented
+// by types that are capable of encoding and decoding collections keys.
+type KeyCodec[T any] interface {
+	// EncodeKey writes the key bytes into the buffer. Returns the number of
+	// bytes written. The implementer must expect the buffer to be at least
+	// of length equal to Size(K) for all encodings.
+	// It must also return the number of written bytes which must be
+	// equal to Size(K) for all encodings not involving varints.
+	// In case of encodings involving varints then the returned
+	// number of written bytes is allowed to be smaller than Size(K).
+	EncodeKey(buffer []byte, key T) (int, error)
+	// DecodeKey reads from the provided bytes buffer to decode
+	// the key T. Returns the number of bytes read, the type T
+	// or an error in case of decoding failure.
+	DecodeKey(buffer []byte) (int, T, error)
```

Comment:
For consistency ;)
```suggestion
	Encode(buffer []byte, key T) (int, error)
	// Decode reads from the provided bytes buffer to decode
	// the key T. Returns the number of bytes read, the type T
	// or an error in case of decoding failure.
	Decode(buffer []byte) (int, T, error)
```